torchbenchmark/models/nvidia_deeprecommender/nvtrain.py

```diff
@@ -0,0 +1,403 @@
+      self.rencoder = self.rencoder.cuda()
+  
+    if self.args.optimizer == "adam":
+      self.optimizer = optim.Adam(self.rencoder.parameters(),
+                                  lr=self.args.lr,
+                                  weight_decay=self.args.weight_decay)
+    elif self.args.optimizer == "adagrad":
+      self.optimizer = optim.Adagrad(self.rencoder.parameters(),
+                                lr=self.args.lr,
+                                weight_decay=self.args.weight_decay)
+    elif self.args.optimizer == "momentum":
+      self.optimizer = optim.SGD(self.rencoder.parameters(),
+                            lr=self.args.lr, momentum=0.9,
+                            weight_decay=self.args.weight_decay)
+      self.scheduler = MultiStepLR(self.optimizer, milestones=[24, 36, 48, 66, 72], gamma=0.5)
+    elif args.optimizer == "rmsprop":
+      self.optimizer = optim.RMSprop(self.rencoder.parameters(),
+                                lr=self.args.lr, momentum=0.9,
+                                weight_decay=self.args.weight_decay)
+    else:
+      raise  ValueError('Unknown optimizer kind')
+  
+    self.t_loss = 0.0
+    self.t_loss_denom = 0.0
+    self.denom = 0.0
+    self.total_epoch_loss = 0.0
+    self.global_step = 0
+  
+    if self.args.noise_prob > 0.0:
+      self.dp = nn.Dropout(p=self.args.noise_prob)
+
+  def DoTrain(self):
+  
+    self.rencoder.train()
+    #if self.args.optimizer == "momentum":
+    #  self.scheduler.step()
+  
+    for i, mb in enumerate(self.data_layer.iterate_one_epoch()):
+  
+      inputs = Variable(mb.cuda().to_dense() if self.args.use_cuda else mb.to_dense())
+      self.optimizer.zero_grad()
+  
+      outputs = self.rencoder(inputs)
+  
+      loss, num_ratings = model.MSEloss(outputs, inputs)
+      loss = loss / num_ratings
+      loss.backward()
+      self.optimizer.step()
+      self.global_step += 1
+      self.t_loss += loss.item()
+      self.t_loss_denom += 1
+  
+      if not self.args.nooutput:
+        if i % self.args.summary_frequency == 0:
+          print('[%d, %5d] RMSE: %.7f' % (self.epoch, i, sqrt(self.t_loss / self.t_loss_denom)))
+          self.logger.scalar_summary("Training_RMSE", sqrt(self.t_loss/self.t_loss_denom), self.global_step)
+          self.t_loss = 0
+          self.t_loss_denom = 0.0
+          log_var_and_grad_summaries(self.logger, self.rencoder.encode_w, self.global_step, "Encode_W")
+          log_var_and_grad_summaries(self.logger, self.rencoder.encode_b, self.global_step, "Encode_b")
+          if not self.rencoder.is_constrained:
+            log_var_and_grad_summaries(self.logger, self.rencoder.decode_w, self.global_step, "Decode_W")
+          log_var_and_grad_summaries(self.logger, self.rencoder.decode_b, self.global_step, "Decode_b")
+  
+      self.total_epoch_loss += loss.item()
+      self.denom += 1
+  
+      #if args.aug_step > 0 and i % args.aug_step == 0 and i > 0:
+      if self.args.aug_step > 0:
```

Comment:
-this block can also be omitted.  it's just adding random noise to the same input aug_step times, and doing another training step for each of those samples.  we can just benchmark the time to do one step.  you'll note this block is a copypaste of the block just below which also computes model output, loss, and steps the optimizer.